File: leetcode/981.time-based-key-value-store.py
# @leet start
class TimeMap:

    def __init__(self):
        self.map: dict[str, list[tuple[int, str]]] = {}

    def set(self, key: str, value: str, timestamp: int) -> None:
        if key not in self.map:
            self.map[key] = []

        self.map[key].append((timestamp, value))

    def get(self, key: str, timestamp: int) -> str:
        if key not in self.map:
            return ""

        arr = self.map[key]

        left = 0
        right = len(arr) - 1

        while left <= right:
            mid = (left + right) // 2

            if arr[mid][0] <= timestamp:
                left = mid + 1
            else:
                right = mid - 1

        return arr[right][1] if right != -1 else ""


# Your TimeMap object will be instantiated and called as such:
# obj = TimeMap()
# obj.set(key,value,timestamp)
# param_2 = obj.get(key,timestamp)
# @leet end

File: leetcode/981.time-based-key-value-store.js
// @leet start
var TimeMap = function () {
  this.map = new Map(); // Map<string, [int, string][]>
};

/**
 * @param {string} key
 * @param {string} value
 * @param {number} timestamp
 * @return {void}
 */
TimeMap.prototype.set = function (key, value, timestamp) {
  if (!this.map.has(key)) {
    this.map.set(key, []);
  }

  this.map.get(key).push([timestamp, value]);
};

/**
 * @param {string} key
 * @param {number} timestamp
 * @return {string}
 */
TimeMap.prototype.get = function (key, timestamp) {
  if (!this.map.has(key)) {
    return "";
  }

  const vals = this.map.get(key);

  if (vals[0][0] > timestamp) {
    return "";
  }

  let left = 0;
  let right = vals.length - 1;

  while (left <= right) {
    const mid = Math.floor((left + right) / 2);

    if (vals[mid][0] <= timestamp) {
      left = mid + 1;
    } else {
      right = mid - 1;
    }
  }

  return vals[right][1];
};

/**
 * Your TimeMap object will be instantiated and called as such:
 * var obj = new TimeMap()
 * obj.set(key,value,timestamp)
 * var param_2 = obj.get(key,timestamp)
 */
// @leet end
